### PR TITLE
#12486 and #49 from alpha-beta-issues fixes

### LINKED ISSUE #12486: Source Google Ads Sync option to skip inactive accounts rather than failing sync for all accounts
## Tell us about the problem you're trying to solve

When using the Customer ID(s) feature in Google Ads source, an account may go inactive if it has not been used after some time. This causes the sync to fail for all accounts until the inactive account id is removed from the source configuration. 

## Describe the solution you’d like

Preferred solutions, as notifications about an account expiration happen from Google already:
* Raise a warning instead of an error when the error_code from Google is "authorization_error: CUSTOMER_NOT_ENABLED" is thrown, skipping the inactive account. 
* Have an option in the configuration to skip inactive accounts (they may become active again later)

More involved solution:
* Have an option to pull from all active accounts, rather than a manual entry of individual accounts
* Similar to selecting the streams in the connection, have the ability to select from accounts and possibly see the status

Another thought:
* Scan all accounts in the Source setup for being active. If the Airbyte API allows at a later time to test connection then this can be checked before any syncs are started via an orchestrator using the API.

## Describe the alternative you’ve considered or used
* Remove inactive accounts manually when an error occurs
* Create a source for each individual account

## Additional context

Section from the error log:
```
google.ads.googleads.errors.GoogleAdsException: (<_InactiveRpcError of RPC that terminated with:
	status = StatusCode.PERMISSION_DENIED
	details = "The caller does not have permission"
	debug_error_string = "{"created":"@1651243103.611933974","description":"Error received from peer ipv4:000.000.000.000:443","file":"src/core/lib/surface/call.cc","file_line":903,"grpc_message":"The caller does not have permission","grpc_status":7}"
>, <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.PERMISSION_DENIED
	details = "The caller does not have permission"
	debug_error_string = "{"created":"@1651243103.611933974","description":"Error received from peer ipv4:000.000.000.000:443","file":"src/core/lib/surface/call.cc","file_line":903,"grpc_message":"The caller does not have permission","grpc_status":7}"
>, errors {
  error_code {
    authorization_error: CUSTOMER_NOT_ENABLED
  }
  message: "The customer can\'t be used because it isn\'t enabled."
}
```

## Are you willing to submit a PR?
Not at this time

### PR DESCRIPTION
**Pull request by davydov-d**
## What
[Source Google Ads Sync option to skip inactive accounts rather than failing sync for all accounts](https://github.com/airbytehq/airbyte/issues/12486):
 - Since the standard `Accounts` stream was used to fetch the customers info, it used a `where` clause in its GAQL query with a `segment.date` filter which filtered out all the customers
  - Even more, only the first customer account among those that could be fetched(if any) was checked for being a manager
  - If that account was not a manager, all the streams were instantiated with a full `customer_ids` list, despite of that some of them could be managers which would lead to errors further when reading

[Source: Google Ads - with GAQL not able to create connection](https://github.com/airbytehq/alpha-beta-issues/issues/49):
 - After making the request for validating the GAQL query to Google Ads API, we get the generator, which would raise an exception in case of error only when being iterated over. Which wasn't done
 - A bug in formatting the custom query: some keyword patterns missed whitespaces, what led to making an invalid query (`parameters` in regex instead of `_parameters_` made the function insert `where` clause right inside the `select` clause)

Other bug fixes:
 - `end_date` param was partially ignored
## How
 - Added another service stream, not exposed to a user, that would fetch the customers by their IDs independent of date segments
- Refactored the code to check `is_manager` attribute of each customer, not only the first one.
- Instantiate the streams depending on the `customer.is_manager` field only with non-manager accounts as params
- Try-catch auth and query errors and log them when reading records and checking the connection instead of failing the stream for all customer accounts
- Fixed the bug in the custom query formatter (added whitespaces so that partial word matches would not be treated as key words)
- Added iteration over the response when validating the custom query so that API errors get raised instead of being ignored
- Updated all the imports and links from API v8 to API v9
- Fixed issue with the `end_date` parameter 
- Updated tests

### CODE REVIEW

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/connectors/source-google-ads/integration_tests/test_incremental.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def test_incremental_sync(config, configured_catalog):
             AirbyteLogger(),
             config,
             ConfiguredAirbyteCatalog.parse_obj(configured_catalog),
-            {"ad_group_ad_report": {"segments.date": latest_state}},
+            {"ad_group_ad_report": {config["customer_id"]: {"segments.date": latest_state}}},
```
Comment (evantahler) [inline airbyte-integrations/connectors/source-google-ads/integration_tests/test_incremental.py:61]: With a change to the state format, what will happen to existing incremental syncs when users upgrade?  Will they start over and perhaps see duplicate data?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py [hunk 2]
```diff
@@ -92,42 +82,42 @@ def is_metrics_in_custom_query(query: str) -> bool:
...
-                    return False, f"Custom query should not contain {CustomQuery.cursor_field}"
-
-                req_q = CustomQuery.insert_segments_date_expr(query, "1980-01-01", "1980-01-01")
-                for customer_id in google_api.customer_ids:
-                    google_api.send_request(req_q, customer_id=customer_id)
+            for customer in customers:
+                for query in config.get("custom_queries", []):
+                    query = query.get("query")
+                    if customer.is_manager_account and self.is_metrics_in_custom_query(query):
+                        return False, f"Metrics are not available for manager account. Check fields in your custom query: {query}"
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:94]:
```suggestion
                        return False, f"Metrics are not available for manager account {customer.id}. Please remove metrics fields in your custom query: {query}."
```

Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:94]: we should put a note in the docs that mentions that custom queries must work for ALL account ids input. If i sync 9 non-manager accountss, and 1 manager account, the connection check will fail. Alternatively we can just skip that custom stream for the manager account and log a message to the user.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py [hunk 3]
```diff
@@ -95,42 +99,64 @@ def parse_response(self, response: SearchPager) -> Iterable[Mapping]:
...
     def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
-        for customer_id in self.google_ads_client.customer_ids:
-            self._customer_id = customer_id
-            yield {}
+        for customer in self.customers:
+            yield {"customer_id": customer.id}
 
     def read_records(self, sync_mode, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
-        if not stream_slice:
+        if stream_slice is None:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:106]: why this change? when would a slice ever be none?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py [hunk 4]
```diff
@@ -95,42 +99,64 @@ def parse_response(self, response: SearchPager) -> Iterable[Mapping]:
...
+        customer_id = stream_slice["customer_id"]
+        response_records = self.google_ads_client.send_request(self.get_query(stream_slice), customer_id=customer_id)
+        try:
+            for response in response_records:
+                yield from self.parse_response(response)
+        except GoogleAdsException as exc:
+            if not self.CATCH_API_ERRORS:
+                raise
+            for error in exc.failure.errors:
+                if error.error_code.authentication_error or error.error_code.authorization_error or error.error_code.query_error:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:118]: why not check for the specific issue of an account being inactive instead of ignoring all authn/authz & query errors? (why are we ignoring query errors btw?) 
it seems this way would cause us to ignore actual problems

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py (hunk 2 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py]: > Alternatively we can just skip that custom stream for the manager account and log a message to the user.
Done

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 4 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py]: Agree. Fixed

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py (hunk 2 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py]: Fixed

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 3 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:107]: when the input state contains abnormally large value - the slice is None and no records are emitted

File: airbyte-integrations/connectors/source-google-ads/integration_tests/test_incremental.py (hunk 1 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-google-ads/integration_tests/test_incremental.py:61]: No, nothing out of the ordinary should happen. The connector supports old state format as well as the new one. Whenever it updates the state value, it is done according to the new format. This commit does not actually change state format (just moved logic from deprecated `get_updated_state` to `state` [property](https://github.com/airbytehq/airbyte/pull/12914/files#diff-3f54675c5be5786da7f7ad049d4216b6a2bd4490dd42e732615fbdfde12be1d5R149)). 

The change in tests is made because the connector emits exactly the same state message it got as an input whenever it is abnormally large, without updating the state format.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py (hunk 2 above)
Comment (davydov-d) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py]: Done https://github.com/airbytehq/airbyte/pull/12914/files#diff-04615dae4ed388b7d6ab2905249f18c69c801d1ab6de1d3d27bf2030491b7db5R115

**Review by sherifnada — APPROVED** [review]

**Review by evantahler — APPROVED** [review]
